### ISSUE TITLE
Consecutive Transformer trainings in same process are non-deterministic at identical seed

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1383]
## Summary

Two back-to-back trainings of \`Transformer<float>\` with **identical** seed, **identical** \`architecture.RandomSeed\`, **identical** training data, **identical** loss + optimizer, on the **same engine** (verified on \`CpuEngine\`), within the **same process**, produce **different** post-training weights.

Reproducer added in this PR at \`tests/AiDotNet.Tests/IntegrationTests/Engines/GpuTransformerDeterminismTests.cs\`.

## Reproducer

```csharp
[Fact]
public void Transformer_Train_TwoRunsAtSameSeed_ProduceIdenticalWeights()
{
    var (arch, xTrain, yTrain) = BuildFixture();  // randomSeed: 42 baked in
    double l2_a = TrainOnceAndCaptureL2Norm(arch, xTrain, yTrain);
    double l2_b = TrainOnceAndCaptureL2Norm(arch, xTrain, yTrain);
    Assert.Equal(l2_a, l2_b);  // FAILS
}
```

Observed (V=16, dModel=32, L=1, heads=2, 64 samples, 10 epochs):

```
Engine: CpuEngine
Run A trained-parameter L2 norm: 15.041524156144305
Run B trained-parameter L2 norm: 14.961553497011428
|A - B| = 0.0799707
```

## Why this is a real bug, not benign FP-associativity noise

- FP-associativity reorder noise on parallel reductions for a ~12k-parameter model typically produces L2 deltas in the 1e-6 .. 1e-4 range. **0.08 is 1000× above that floor.**
- HarmonicEngine's byte-LM consumer reproducer (\`Phase_PAPER_A_PathB_Int8_Sanity\`) sees this compound into wildly different convergence trajectories: FP32 top-1 oscillates 0% / 3% / 5% / 6.5% / 19% across runs at identical seed (varies by engine + process boundary + prior in-process training history).
- The deltas reproduce on \`CpuEngine\` within the same process — so this is **not** GPU-specific FP non-determinism.

## Suspected root cause

A downstream RNG consumer in the training path reads from \`RandomHelper.ThreadSafeRandom\`. That helper's per-thread \`Random\` is counter-seeded at thread first-touch (deterministic for thread ID 1, 2, 3, ...), but the Random's STATE advances cumulatively as it's consumed. So Run A advances thread T's Random by some amount, and Run B's identical training starts from where Run A left off — not from the same initial RNG state.

Candidates worth grepping:
- \`AdamOptimizer\` / batch shuffle code
- \`Dropout\` / regularization layers (the model in the reproducer has \`dropoutRate: default\` — and the default Vaswani schedule includes dropout)
- Lazy parameter init in \`MultiHeadAttentionLayer\` / \`DenseLayer\` (if any layer fall-through to \`LayerBase.Random\` after \`architecture.RandomSeed\` is consumed)

## Why this matters

This is the root cause behind the seemingly intermittent test failures HarmonicEngine has been chasing as "GPU non-determinism" or "FP non-associativity noise". The bug is more fundamental — every consumer that runs multiple trainings in one process (parameter sweeps, multi-seed averages, A/B arm comparisons inside one test) gets a different trained model on each invocation even though the seed parameter is bit-identical.

## Expected behavior

\`Random R = RandomHelper.ThreadSafeRandom\` should either:
- (a) be a **fresh** Random per training session (callers explicitly seed via \`CreateSeededRandom(seed)\`); OR
- (b) be reset to its initial seed-derived state at training-session boundaries; OR
- (c) all training-path RNG consumers explicitly use \`RandomHelper.CreateSeededRandom(architecture.RandomSeed.Value)\` rather than the shared thread-local one.

(c) is the cleanest fix because it makes the determinism contract explicit on each consumer.

## Related issues

- \`#1380\` — original BuildAsync byte-LM mode collapse (fixed in this same PR by Vector→Tensor bridge in CalculateGradient)
- \`#1382\` — TransformerArchitecture(layers:, numEncoderLayers > 0) silent-drop (fixed in this same PR by ctor validation)
- This issue — third determinism bug in the same family, surfaced once the first two were fixed
